bpo-38811: Check for presence of os.link method in pathlib

tohojo opened the pull request:
Commit 6b5b013bcc22 ("[bpo-26978](https://bugs.python.org/issue26978): Implement pathlib.Path.link_to (Using
os.link) (GH-12990)") introduced a new link_to method in pathlib. However,
this makes pathlib crash when the 'os' module is missing a 'link' method.

Fix this by checking for the presence of the 'link' method on pathlib
module import, and if it's not present, turn it into a runtime error like
those emitted when there is no lchmod() or symlink().

Signed-off-by: Toke Høiland-Jørgensen <toke@redhat.com>

<!-- issue-number: [bpo-38811](https://bugs.python.org/issue38811) -->
https://bugs.python.org/issue38811
<!-- /issue-number -->

tohojo committed:
bpo-38811: Check for presence of os.link method in pathlib

Commit 6b5b013bcc22 ("bpo-26978: Implement pathlib.Path.link_to (Using
os.link) (GH-12990)") introduced a new link_to method in pathlib. However,
this makes pathlib crash when the 'os' module is missing a 'link' method.

Fix this by checking for the presence of the 'link' method on pathlib
module import, and if it's not present, turn it into a runtime error like
those emitted when there is no lchmod() or symlink().

Signed-off-by: Toke Høiland-Jørgensen <toke@redhat.com>

cool-RR reviewed (approved):
Looks good to me.

brandtbucher commented on Lib/pathlib.py:
```diff
@@ -418,7 +418,11 @@ def lchmod(self, pathobj, mode):
+        link_to = os.link
+    else:
+        def link_to(self, target):
+            raise NotImplementedError("link() not available on this system")
```
```suggestion
            raise NotImplementedError("os.link() not available on this system")
```

brandtbucher commented on Lib/test/test_pathlib.py:
```diff
@@ -1778,6 +1778,15 @@ def test_link_to(self):
         self.assertEqual(os.stat(r).st_size, size)
         self.assertTrue(q.stat)
 
+    @unittest.skipIf(hasattr(os, "link"), "the os module link method is present")
```
```suggestion
    @unittest.skipIf(hasattr(os, "link"), "os.link() is present")
```

brandtbucher commented on Misc/NEWS.d/next/Library/2019-11-15-18-06-04.bpo-38811.AmdQ6M.rst:
```diff
@@ -0,0 +1 @@
+Fix crash in pathlib when 'os' module is missing a 'link' method.
```
```suggestion
Fix an unhandled exception in :mod:`pathlib` when :meth:`os.link` is missing.
```

brandtbucher reviewed (changes requested):
Looks good, just a bit of cleanup here:

serhiy-storchaka commented on Lib/pathlib.py:
```diff
@@ -418,7 +418,11 @@ def lchmod(self, pathobj, mode):
+    if hasattr(os, "link"):
+        link_to = os.link
+    else:
+        def link_to(self, target):
```
`@staticmethod`

tohojo replied:
Hmm, figured that didn't really matter much since the function is not doing anything with its args anyway. But OK, will fix :)

serhiy-storchaka replied:
It matters. Without this you will get an error:
```
======================================================================
ERROR: test_link_to_failure (test.test_pathlib.PathTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/serhiy/py/cpython/Lib/test/test_pathlib.py", line 1788, in test_link_to_failure
    p.link_to(q)
  File "/home/serhiy/py/cpython/Lib/pathlib.py", line 1364, in link_to
    self._accessor.link_to(self, target)
TypeError: link_to() takes 2 positional arguments but 3 were given

======================================================================
ERROR: test_link_to_failure (test.test_pathlib.PosixPathTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/serhiy/py/cpython/Lib/test/test_pathlib.py", line 1788, in test_link_to_failure
    p.link_to(q)
  File "/home/serhiy/py/cpython/Lib/pathlib.py", line 1364, in link_to
    self._accessor.link_to(self, target)
TypeError: link_to() takes 2 positional arguments but 3 were given

----------------------------------------------------------------------
```

tohojo replied:
OIC; didn't realise that was how the function was being called. Right, fixed (properly on the second try) :)

serhiy-storchaka commented on Lib/test/test_pathlib.py:
```diff
@@ -1778,6 +1778,15 @@ def test_link_to(self):
         self.assertTrue(q.stat)
 
+    @unittest.skipIf(hasattr(os, "link"), "os.link() is present")
+    def test_link_to_failure(self):
```
Would not `test_link_to_not_implemented` be better name?

tohojo replied:
Fixed, and added :)

serhiy-storchaka reviewed (commented):
`test_link_to` fails if `os.link` is not available. Could you please add `skipIf(hasattr(os, "link"))` to it?

Would be nice to add also a test for `Path.symlink` in case when `os.symlink` is not available.

tohojo committed:
Lib: Add pathlib test for missing symlink support

This adds a test case to test_pathlib.py to test the NotImplementedError
raised when symlink support is not present on the system. Also add the
missing @staticmethod decorator to the function definition in pathlib.py
for that error path.

Signed-off-by: Toke Høiland-Jørgensen <toke@redhat.com>

serhiy-storchaka commented on Misc/NEWS.d/next/Library/2019-11-15-18-06-04.bpo-38811.AmdQ6M.rst:
```diff
@@ -0,0 +1 @@
+Fix an unhandled exception in :mod:`pathlib` when :meth:`os.link` is missing.
```
```suggestion
Fix an unhandled exception in :mod:`pathlib` when :meth:`os.link` is missing. Patch by Toke Høiland-Jørgensen.
```

tohojo replied:
Fine with me, obviously :)
